grpc/stream.go

```diff
@@ -334,15 +337,13 @@ func (s *stream) closeWithError(err error) error {
 	return s.callErrorListeners(err)
 }
 
-// close changes the stream state to closed and triggers the end event listeners
+// close closes the stream and call end event listeners
+// Note: in the regular closing the io.EOF could come
 func (s *stream) close(err error) {
-	if s.state == closed {
+	if err == nil {
```

Comment:
Checking for the state here will give us less since it's is a writing state, but here, we interested in the form of the stream in general. So, either we should split the responsibilities and introduce the writing state and the "just" stream state. But that could be not very clear.

How about using the select/default to check if the channel is closed?

Reply:
Hmm I am not against this for now I guess, so let's keep it.